src/main/java/org/myrobotlab/service/abstracts/AbstractServo.java

```diff
@@ -821,24 +823,36 @@ public void setRest(double rest) {
   @Override
   @Config
   public void setSpeed(Double degreesPerSecond) {
-    if (degreesPerSecond == null) {
-      log.info("disabling speed control");
-      speed = null;
-      broadcast("publishServoSetSpeed", this);
-      broadcastState();
-      return;
-    }
     // KW: TODO: technically the Arduino will read this speed as a 16 bit int..
     // so max Speed is 32,767 ...
     // if (maxSpeed != -1 && degreesPerSecond != null && degreesPerSecond >
     // maxSpeed) {
     // speed = maxSpeed;
     // log.info("Trying to set speed to a value greater than max speed");
     // }
-    speed = degreesPerSecond;
+    if (maxSpeed != null && (degreesPerSecond == null || degreesPerSecond > maxSpeed)) {
+    	speed = maxSpeed;
+    	return;
+    } else {
+    	speed = degreesPerSecond;
+    }
+    
+    if (degreesPerSecond == null) {
+      log.info("disabling speed control");
+    }    
     broadcast("publishServoSetSpeed", this);
     broadcastState();
   }
+  
+  @Override
+  public void setMaxSpeed(Double degreesPerSecond) {
+	 if (degreesPerSecond != null && (speed == null || speed > degreesPerSecond)){
```

Comment:
This method also sets the current speed? That seems wrong to me... If I say setMaxSpeed(100)  I expect the servo will never move faster than that.  I wouldn't expect that the servo would start moving at the max speed unless I explicitly called setSpeed on it..

Reply:
Whether its done here (which I think is cleaner) ..
or if its set in processMove - if you setMaxSpeed(100) and you have "unbounded" speed currently what is your expectation?
"unbounded" is faster than 100
so speed should be 100 in that case ...
Another use case .. what if the speed is currently 200 .. and you setMaxSpeed(100) -  what would be your expectation that its left at 200 ?